File: der_diffusion_tensor.cpp
#include "diffusion_tensor.h"
#include "containers.h"
#include "test_cases.h"
#include "solid_stress_der_der.h"
#include "solid_stress_der.h"
#include "solid_stress.h"
#include "sed_hsf.h"
#include "sed_hsf_der.h"
#include "kronecker.h"
#include "update_utilities.h"

namespace global 
{
	extern double D0; 
    extern int diff_idx;
}

void der_diffusion_tensor(Matrix<double>& u, Block<double> &B, int r)
{
	struct test_cases* pt_test = get_tests();

	int m = size(u, 1); 
	int n = size(u, 2);

	Matrix<double> &delta = *pt_test->delta;
	Matrix<double> *ptphik =  new Matrix<double>(m, 1); Matrix<double> &phik = *ptphik;
	
	double phit, p2, quot, sigedd, siged, sige, wphit, wphitd, theta, psih;

	for (int k = 1; k <= n; k++)
	{
		get_col(phik, u, k);
		phit = sum1D(phik);
		p2 = dot_product(delta, phik);

		if (phit > 1)
    	{
      		printf("\nFor column %d\n", k);
      		throw std::invalid_argument("\nError inside der_diffusion_tensor: phit > 1 !!\n");
    	}

    	if (global::diff_idx != 0)
		{
			quot = 1/phit*(1 - phit);
			sigedd = solid_stress_der_der(phit);
        	siged  = solid_stress_der(phit);
        	sige   = solid_stress(phit);
        	wphit  = sed_hsf(phit);
        	wphitd = sed_hsf_der(phit);

			for (int i = 1; i <= m; i++)
			{
				theta = phik(i)*(delta(i) - p2);
				for (int j = 1; j <= m; j++)
				{
					psih = delta(i)*kronecker(i, j) - delta(j)*phik(i) - theta/phit;
                	B(k, i, j) = pt_test->mu_g*((phit*wphitd - wphit)*(theta*siged - psih*sige*quot)/pow(phit, 2) 
							   + wphit*(theta*sigedd - (theta*sige/pow(phit, 2) + psih*((1 - phit)*siged + sige)/(1 - phit))/(1 - phit))/phit);					
				}			
			}
		}
		else
		{
			for (int i = 1; i <= m; i++) 
				B(k, i, i) = -global::D0*pt_test->nexp*pow(1 - phit, pt_test->nexp - 1);
		}
	}
	
	delete ptphik;
	delete pt_test->delta;
	delete pt_test;
}

File: der_diffusion_tensor.m
function B=der_diffusion_tensor(u, B, r)

%int:n,m,i,j,k,r
%double:phit,phii,D0,nexp
%vector:u
%matrix:u,B

global D0 nexp diff_idx delta mu_g;

n = size(u, 2);
m = size(u, 1);

phik = zeros(m, 1);
for k = 1:n
    
    phit = 0;
    p2 = 0;
    for i = 1:m
        phii = u(i, k);
        phit = phit + phii;
        p2 = p2 + phii*delta(i);
        phik(i) = phii;
    end

    if (phit > 1)
        fprintf("\nphit > 1 in the column %d!!\n\n", k);
        error('phit > 1!!');
    end
    
    if (diff_idx ~= 0)
        quot   = 1/phit*(1 - phit);
        sigedd = solid_stress_der_der(phit);
        siged  = solid_stress_der(phit);
        sige   = solid_stress(phit);
        wphit  = sed_hsf(phit);
        wphitd = sed_hsf_der(phit);
        for i = 1:m
            theta = phik(i)*(delta(i) - p2);
            for j = 1:m
                psih = delta(i)*kronecker(i, j) - delta(j)*phik(i) - theta/phit;
                B(i, j, k) = mu_g*((phit*wphitd - wphit)*(theta*siged - psih*sige*quot)/phit^2 + ... 
                             wphit*(theta*sigedd - (theta*sige/phit^2 + psih*((1 - phit)*siged + sige)/(1 - phit))/(1 - phit))/phit);
            end
        end
    else
        for i=1:m
            B(i, i, k) = -D0*nexp*(1-phit)^(nexp-1);
        end
    end
end